File: frontend/src/services/notifications.ts
// Frontend service for managing notifications

export interface Notification {
  id: string;
  title: string;
  message: string;
  type: string;
  read: boolean;
  createdAt: string;
  data?: Record<string, any>;
}

class NotificationService {
  private baseUrl = "/api/v1/notifications";
  
  // Subscribe to notification types
  async subscribeToNotifications(notificationTypes: string[]): Promise<{ success: boolean; message: string }> {
    const response = await fetch(`${this.baseUrl}/subscribe`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${localStorage.getItem('accessToken')}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ notification_types: notificationTypes })
    });
    
    if (!response.ok) {
      throw new Error(`Failed to subscribe to notifications: ${response.statusText}`);
    }
    
    return response.json();
  }
  
  // Unsubscribe from notification types
  async unsubscribeFromNotifications(notificationTypes: string[]): Promise<{ success: boolean; message: string }> {
    const response = await fetch(`${this.baseUrl}/unsubscribe`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${localStorage.getItem('accessToken')}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ notification_types: notificationTypes })
    });
    
    if (!response.ok) {
      throw new Error(`Failed to unsubscribe from notifications: ${response.statusText}`);
    }
    
    return response.json();
  }
  
  // Get user's notifications
  async getUserNotifications(): Promise<{ notifications: Notification[] }> {
    const response = await fetch(`${this.baseUrl}`, {
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${localStorage.getItem('accessToken')}`,
        'Content-Type': 'application/json'
      }
    });
    
    if (!response.ok) {
      throw new Error(`Failed to get notifications: ${response.statusText}`);
    }
    
    return response.json();
  }
  
  // Mark notification as read
  async markNotificationAsRead(notificationId: string): Promise<{ success: boolean; message: string }> {
    const response = await fetch(`${this.baseUrl}/${notificationId}/read`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${localStorage.getItem('accessToken')}`,
        'Content-Type': 'application/json'
      }
    });
    
    if (!response.ok) {
      throw new Error(`Failed to mark notification as read: ${response.statusText}`);
    }
    
    return response.json();
  }
  
  // Send test notification
  async sendTestNotification(
    notificationType: string,
    title: string,
    message: string,
    data?: Record<string, any>
  ): Promise<{ success: boolean; message: string }> {
    const response = await fetch(`${this.baseUrl}/test`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${localStorage.getItem('accessToken')}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        notification_type: notificationType,
        title,
        message,
        data
      })
    });
    
    if (!response.ok) {
      throw new Error(`Failed to send test notification: ${response.statusText}`);
    }
    
    return response.json();
  }
  
  // Request notification permissions (for browser notifications)
  async requestNotificationPermission(): Promise<boolean> {
    if (!('Notification' in window)) {
      // Browser doesn't support desktop notifications
      return false;
    }
    
    if (Notification.permission === 'granted') {
      return true;
    }
    
    if (Notification.permission !== 'denied') {
      const permission = await Notification.requestPermission();
      return permission === 'granted';
    }
    
    return false;
  }
  
  // Show browser notification
  async showBrowserNotification(title: string, message: string, icon?: string): Promise<void> {
    const hasPermission = await this.requestNotificationPermission();
    if (hasPermission) {
      new Notification(title, {
        body: message,
        icon: icon || '/favicon.ico'
      });
    }
  }
}

// Global instance
export const notificationService = new NotificationService();

File: api/routers/notifications.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession  # FIX: Async session
from samokoder.core.db.session import get_async_db  # FIX: Async DB
from samokoder.core.db.models.user import User
from samokoder.core.services.notification_service import notification_service
from samokoder.api.routers.auth import get_current_user
from typing import List, Optional
import json

router = APIRouter()

@router.post("/notifications/subscribe")
async def subscribe_to_notifications(
    notification_types: List[str],
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)  # FIX: AsyncSession
):
    """
    Subscribe user to notification types
    
    :param notification_types: List of notification types to subscribe to
    :param user: Current user
    :param db: Database session
    :return: Subscription result
    """
    try:
        await notification_service.subscribe_user(user.id, notification_types)
        return {
            "success": True,
            "message": f"Subscribed to notification types: {', '.join(notification_types)}"
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error subscribing to notifications: {str(e)}")


@router.post("/notifications/unsubscribe")
async def unsubscribe_from_notifications(
    notification_types: List[str],
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)  # FIX: AsyncSession
):
    """
    Unsubscribe user from notification types
    
    :param notification_types: List of notification types to unsubscribe from
    :param user: Current user
    :param db: Database session
    :return: Unsubscription result
    """
    try:
        await notification_service.unsubscribe_user(user.id, notification_types)
        return {
            "success": True,
            "message": f"Unsubscribed from notification types: {', '.join(notification_types)}"
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error unsubscribing from notifications: {str(e)}")


@router.get("/notifications")
async def get_user_notifications(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)  # FIX: AsyncSession
):
    """
    Get user's notifications
    
    :param user: Current user
    :param db: Database session
    :return: User's notifications
    """
    try:
        notifications = await notification_service.get_user_notifications(user.id)
        return {
            "notifications": notifications
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error getting notifications: {str(e)}")


@router.post("/notifications/{notification_id}/read")
async def mark_notification_as_read(
    notification_id: str,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)  # FIX: AsyncSession
):
    """
    Mark a notification as read
    
    :param notification_id: Notification ID
    :param user: Current user
    :param db: Database session
    :return: Result
    """
    try:
        success = await notification_service.mark_notification_as_read(user.id, notification_id)
        return {
            "success": success,
            "message": "Notification marked as read" if success else "Failed to mark notification as read"
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error marking notification as read: {str(e)}")


@router.post("/notifications/test")
async def send_test_notification(
    notification_type: str,
    title: str,
    message: str,
    data: Optional[dict] = None,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_async_db)  # FIX: AsyncSession
):
    """
    Send a test notification to the user
    
    :param notification_type: Type of notification
    :param title: Notification title
    :param message: Notification message
    :param data: Additional data
    :param user: Current user
    :param db: Database session
    :return: Result
    """
    try:
        await notification_service.send_notification(
            user.id, 
            notification_type, 
            title, 
            message, 
            data
        )
        return {
            "success": True,
            "message": f"Test notification sent: {title}"
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error sending test notification: {str(e)}")